File: exercises/capstone/ex04_permissions/ex04_permissions.cpp
// Exercise 04 -- the clearance and the permission set (Step 4).
//
// Read exercises/capstone/ex04_permissions/README.md first.
#include "planning/airport/taxi_planner.hpp"

namespace planning::airport {

PermissionSet buildPermissions(const TaxiGraph& gated, const ZoneLayer& layer,
                               const Clearance& clearance) {
  // TODO(you): turn what ATC said into what the search is allowed to do.
  //
  //   routeLabels          straight from clearance.route
  //   enterableRunways     the crossings, plus the destination *only* when the
  //                        clearance says we may enter it
  //   authorizedGates      every hard gate whose `protects` names one of them
  //   mandatoryStops       every other hard gate that has a painted line
  //   goal                 where the mission ends
  //
  // For a runway destination the goal is the holding position on the last
  // taxiway of the cleared route -- or, with a line-up clearance, the runway
  // edge beyond it.  For a stand it is the parking position itself: the end of
  // the stand lead-in line that is inside the stand area.
  //
  // Any hard gate not in authorizedGates gets infinite cost for this mission.
  // The clearance is the only thing in the whole system that can open one.
  (void)gated;
  (void)layer;
  (void)clearance;
  return PermissionSet{};
}

}  // namespace planning::airport

File: src/capstone/problems/ex04-permissions.ts
// Exercise 04 -- the clearance and the permission set (Step 4).
// Brief: docs/capstone/ex04-permissions.md

import {
  type Clearance,
  emptyPermissions,
  type PermissionSet,
  type TaxiGraph,
  type ZoneLayer,
} from '../../airport/index.js';
import { zoneAt } from './ex01-zone-layer.js';

export function buildPermissions(
  gated: TaxiGraph,
  layer: ZoneLayer,
  clearance: Clearance,
): PermissionSet {
  // TODO(you): the clearance is the only thing in the whole system that can
  // open a runway gate.  Turn the sentence into the permission set:
  //
  //   routeLabels          the via list, in order, untouched
  //   enterableRunways     the crossings -- plus the destination, but ONLY
  //                        when clearedToEnterDestination.  "Hold short of 27"
  //                        and "line up 27" are the same runway and opposite
  //                        permissions.
  //   authorizedGates      every hard gate whose `protects` names an enterable
  //                        runway.  Every other hard gate is shut.
  //   mandatoryStops       the unauthorized hard gates that have a painted
  //                        holding position.  A runway *edge* is a hard gate
  //                        too, but it is not somewhere you stop.
  //   goal                 for a runway destination: the hard gate protecting
  //                        it that is also on the last cleared taxiway -- the
  //                        holding position when holding short, the runway edge
  //                        when cleared to enter.  For a stand: the far end of
  //                        that stand's lead-in line.
  //   detail               say what happened, either way.  When there is no
  //                        goal this string is the whole refusal.
  void gated;
  void layer;
  void clearance;
  void zoneAt;
  return emptyPermissions();
}
